File: core/src/main/java/nl/t64/game/rpg/components/party/SkillContainer.java
package nl.t64.game.rpg.components.party;

import com.fasterxml.jackson.annotation.JsonCreator;

import java.beans.ConstructorProperties;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;


class SkillContainer {

    private static final int NUMBER_OF_SKILL_SLOTS = 18;

    private final Map<String, SkillItem> skills;

    private SkillContainer() {
        this.skills = new HashMap<>(NUMBER_OF_SKILL_SLOTS);
    }

    @JsonCreator
    @ConstructorProperties({
            "alchemist", "diplomat", "healer", "loremaster", "mechanic", "merchant",
            "ranger", "stealth", "thief", "troubadour", "warrior", "wizard",
            "hafted", "missile", "pole", "shield", "sword", "thrown"})
    private SkillContainer(int alc, int dip, int hlr, int lor, int mec, int mer,
                           int ran, int stl, int thf, int trb, int war, int wiz,
                           int haf, int mis, int pol, int shd, int swd, int thr) {
        this();
        this.skills.put(SkillType.ALCHEMIST.name(), new Alchemist(alc));
        this.skills.put(SkillType.DIPLOMAT.name(), new Diplomat(dip));
        this.skills.put(SkillType.HEALER.name(), new Healer(hlr));
        this.skills.put(SkillType.LOREMASTER.name(), new Loremaster(lor));
        this.skills.put(SkillType.MECHANIC.name(), new Mechanic(mec));
        this.skills.put(SkillType.MERCHANT.name(), new Merchant(mer));
        this.skills.put(SkillType.RANGER.name(), new Ranger(ran));
        this.skills.put(SkillType.STEALTH.name(), new Stealth(stl));
        this.skills.put(SkillType.THIEF.name(), new Thief(thf));
        this.skills.put(SkillType.TROUBADOUR.name(), new Troubadour(trb));
        this.skills.put(SkillType.WARRIOR.name(), new Warrior(war));
        this.skills.put(SkillType.WIZARD.name(), new Wizard(wiz));
        this.skills.put(SkillType.HAFTED.name(), new Hafted(haf));
        this.skills.put(SkillType.MISSILE.name(), new Missile(mis));
        this.skills.put(SkillType.POLE.name(), new Pole(pol));
        this.skills.put(SkillType.SHIELD.name(), new Shield(shd));
        this.skills.put(SkillType.SWORD.name(), new Sword(swd));
        this.skills.put(SkillType.THROWN.name(), new Thrown(thr));
    }

    List<SkillType> getAllAboveZero() {
        return Arrays.stream(SkillType.values())
                     .filter(this::hasPositiveQuantity)
                     .collect(Collectors.toList());
    }

    int getRankOf(SkillType skillType) {
        return skills.get(skillType.name()).rank;
    }

    int getBonusSkillOf(SkillType skillType) {
        return skills.get(skillType.name()).bonus;
    }

    int getCalculatedTotalSkillOf(SkillType skillType) {
        return skills.get(skillType.name()).getCalculatedTotal();
    }

    int getXpCostForNextLevelOf(SkillType skillType) {
        final int totalLoremaster = getCalculatedTotalSkillOf(SkillType.LOREMASTER);
        return skills.get(skillType.name()).getXpCostForNextLevel(totalLoremaster);
    }

    int getGoldCostForNextLevelOf(SkillType skillType) {
        return skills.get(skillType.name()).getGoldCostForNextLevel();
    }

    private boolean hasPositiveQuantity(SkillType skillType) {
        return getRankOf(skillType) > 0;
    }

}

File: core/src/nl/t64/game/rpg/components/party/skills/SkillContainer.kt
package nl.t64.game.rpg.components.party.skills

import java.beans.ConstructorProperties


private const val NUMBER_OF_SKILL_SLOTS = 19

class SkillContainer() {

    private val skills: MutableMap<String, SkillItem> = HashMap(NUMBER_OF_SKILL_SLOTS)

    @ConstructorProperties(
        "alchemist", "diplomat", "healer", "loremaster", "mechanic", "merchant",
        "ranger", "stealth", "thief", "troubadour", "warrior", "wizard",
        "hafted", "missile", "pole", "shield", "sword", "thrown",
        "bite")
    constructor(alc: Int, dip: Int, hlr: Int, lor: Int, mec: Int, mer: Int,
                ran: Int, stl: Int, thf: Int, trb: Int, war: Int, wiz: Int,
                haf: Int, mis: Int, pol: Int, shd: Int, swd: Int, thr: Int,
                bit: Int
    ) : this() {
        this.skills[SkillItemId.ALCHEMIST.name] = Alchemist(alc)
        this.skills[SkillItemId.DIPLOMAT.name] = Diplomat(dip)
        this.skills[SkillItemId.HEALER.name] = Healer(hlr)
        this.skills[SkillItemId.LOREMASTER.name] = Loremaster(lor)
        this.skills[SkillItemId.MECHANIC.name] = Mechanic(mec)
        this.skills[SkillItemId.MERCHANT.name] = Merchant(mer)
        this.skills[SkillItemId.RANGER.name] = Ranger(ran)
        this.skills[SkillItemId.STEALTH.name] = Stealth(stl)
        this.skills[SkillItemId.THIEF.name] = Thief(thf)
        this.skills[SkillItemId.TROUBADOUR.name] = Troubadour(trb)
        this.skills[SkillItemId.WARRIOR.name] = Warrior(war)
        this.skills[SkillItemId.WIZARD.name] = Wizard(wiz)
        this.skills[SkillItemId.HAFTED.name] = Hafted(haf)
        this.skills[SkillItemId.MISSILE.name] = Missile(mis)
        this.skills[SkillItemId.POLE.name] = Pole(pol)
        this.skills[SkillItemId.SHIELD.name] = Shield(shd)
        this.skills[SkillItemId.SWORD.name] = Sword(swd)
        this.skills[SkillItemId.THROWN.name] = Thrown(thr)
        this.skills[SkillItemId.BITE.name] = Bite(bit)
    }


    fun getById(skillItemId: SkillItemId): SkillItem {
        return skills[skillItemId.name]!!
    }

    fun getAllAboveZero(): List<SkillItem> {
        return SkillItemId.values()
            .map { skills[it.name]!! }
            .filter { hasPositiveQuantity(it) }
    }

    private fun hasPositiveQuantity(skillItem: SkillItem): Boolean {
        return skillItem.rank > 0
    }

}
